src/radical/entk/appman/wfprocessor.py

```diff
@@ -100,145 +110,173 @@ def _initialize_workflow(self):
+                            exec_stage.parent_pipeline['uid'] = pipe.uid
+                            exec_stage.parent_pipeline['name'] = pipe.name
+                            exec_stage._assign_uid(self._sid)
+
+                        # We need to check both INITIAL and SCHEDULED as
+                        # there maybe a task in the stage that needs to be
+                        # resubmitted.
+                        exec_tasks = list()
+                        if exec_stage.state in [states.INITIAL,
+                                                states.SCHEDULED]:
+
+                            # If its a new stage, update its state
+                            if exec_stage.state == states.INITIAL:
+
+                                local_transition(obj=exec_stage,
+                                                    obj_type='Stage',
+                                                    new_state=states.SCHEDULING,
+                                                    profiler=self._prof,
+                                                    logger=self._logger,
+                                                    reporter=self._report)
+
+                            exec_tasks = exec_stage.tasks
+
+                        for exec_task in exec_tasks:
+
+                            if (exec_task.state == states.INITIAL) or \
+                                ((exec_task.state == states.FAILED) and
+                                    (self._resubmit_failed)):
+
+                                # Set state of Tasks in current Stage
+                                # to SCHEDULING
+                                local_transition(obj=exec_task,
+                                                    obj_type='Task',
+                                                    new_state=states.SCHEDULING,
+                                                    profiler=self._prof,
+                                                    logger=self._logger,
+                                                    reporter=self._report)
+
+                                # Store the tasks from different pipelines
+                                # into our workload list. All tasks will
+                                # be submitted in bulk and their states
+                                # will be updated accordingly
+                                workload.append(exec_task)
+
+                                # We store the stages since the stages the
+                                # above tasks belong to also need to be
+                                # updated. If its a task that failed, the
+                                # stage is already in the correct state
+                                if exec_task.state == states.FAILED:
+                                    continue
+                                if exec_stage not in scheduled_stages:
+                                    scheduled_stages.append(exec_stage)
+
```

Comment:
At this point you have:
```
def ...:
  try:
    while:
      for:
        with:
          for:
            if:
             if:
```
which is hard to keep in mind when reading this code.  One option would be to make the thread main routine short and self contained, like:

```
def _enqueue(self):
  try:
    while not terminated:
      work()
  except:
     oops
```

which keeps the except closer than a gazillion lines to the try ;-)  And also lets `work` focus on the actual semantics.

Reply:
Well, that reduced one level of indentation. As long as the code is within the margins, it works for me.
But seeing that you would prefer them shorter, maybe we can do the following:

```
def _enqueue(self):
  try:
    while not terminated:
      workload = create_workload()
      if workload:
        execute_workload()
  except:
     oops
```

There is a logical separation in what work() does as well and I would find it more meaningful. What do you think?

Reply:
Yes, that works for sure!  Don't take all my comments at face value BTW, I know it's not always possible / sensible to adhere to the guidelines religiously - but your suggestion would make the code more readable IMHO!

Reply:
Cool, will make the change then :+1: